### PR TITLE
cleanup: introduce populateRbdVol(), to fill rbdVol from stage req

### PR DESCRIPTION
**Pull request by humblec**
At present the nodeStageVolume() handle many logic of filling rbdvol
struct based on the request received and this method is complex to
follow. with this patch, filling or populating rbvol struct has been
segregated and handled hence make the stage functions' job easy.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

### CODE REVIEW

File: internal/rbd/nodeserver.go [hunk 1]
```diff
@@ -149,6 +149,65 @@ func healerStageTransaction(ctx context.Context, cr *util.Credentials, volOps *r
...
 
+// populateVolOptions update the fields in rbdVolume struct based on the request it received.
+func (ns *NodeServer) populateVolOptions(
+	ctx context.Context,
+	req *csi.NodeStageVolumeRequest,
+	disableInUseChecks bool,
+	isStaticVol bool,
+	volID string, cr *util.Credentials) (*rbdVolume, error) {
+	var err error
+	volOptions, err := genVolFromVolumeOptions(ctx, req.GetVolumeContext(), req.GetSecrets(), disableInUseChecks)
```
Comment (nixpanic) [inline internal/rbd/nodeserver.go]: `volOptions` is not really a suitable name better call it `rbdVol` or `rv` like done in other places.

File: internal/rbd/nodeserver.go [hunk 2]
```diff
@@ -149,6 +149,65 @@ func healerStageTransaction(ctx context.Context, cr *util.Credentials, volOps *r
 	return nil
 }
 
+// populateVolOptions update the fields in rbdVolume struct based on the request it received.
+func (ns *NodeServer) populateVolOptions(
+	ctx context.Context,
+	req *csi.NodeStageVolumeRequest,
+	disableInUseChecks bool,
+	isStaticVol bool,
```
Comment (nixpanic) [inline internal/rbd/nodeserver.go]: the whole request is passed, is it not possible to detect `isStaticVol` from that?

File: internal/rbd/nodeserver.go [hunk 3]
```diff
@@ -149,6 +149,65 @@ func healerStageTransaction(ctx context.Context, cr *util.Credentials, volOps *r
...
+		err = vi.DecomposeCSIID(volID)
+		if err != nil {
+			err = fmt.Errorf("error decoding volume ID (%s): %w", volID, err)
+
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+
+		j, connErr := volJournal.Connect(volOptions.Monitors, volOptions.RadosNamespace, cr)
+		if connErr != nil {
+			log.ErrorLog(ctx, "failed to establish cluster connection: %v", connErr)
```
Comment (nixpanic) [inline internal/rbd/nodeserver.go]: consider updating `connErr` just like is done for other error cases

File: internal/rbd/nodeserver.go [hunk 4]
```diff
@@ -149,6 +149,65 @@ func healerStageTransaction(ctx context.Context, cr *util.Credentials, volOps *r
 	return nil
 }
 
+// populateVolOptions update the fields in rbdVolume struct based on the request it received.
+func (ns *NodeServer) populateVolOptions(
```
Comment (nixpanic) [inline internal/rbd/nodeserver.go]: Can this function be re-used easier when it isn't a method of `NodeServer`? `ns` does not seem to be used in the function, so tying it to `NodeServer` is sub-optimal.

**Review by nixpanic — DISMISSED** [review]

File: internal/rbd/nodeserver.go (hunk 4 above)
Comment (humblec) [reply internal/rbd/nodeserver.go]: Done. I was more inclined to keep it as  a method though.

File: internal/rbd/nodeserver.go (hunk 3 above)
Comment (humblec) [reply internal/rbd/nodeserver.go]: done.

File: internal/rbd/nodeserver.go (hunk 2 above)
Comment (humblec) [reply internal/rbd/nodeserver.go]: Can be done,  and its done :)

File: internal/rbd/nodeserver.go (hunk 1 above)
Comment (humblec) [reply internal/rbd/nodeserver.go]: changed and kept it with rbdvol

**Comment by humblec** [comment]
@nixpanic comments are addressed. PTAL. Thanks !

File: internal/rbd/nodeserver.go [hunk 5]
```diff
@@ -149,6 +149,65 @@ func healerStageTransaction(ctx context.Context, cr *util.Credentials, volOps *r
 	return nil
 }
 
+// populateVolOptions update the fields in rbdVolume struct based on the request it received.
+func populateRbdVol(
+	ctx context.Context,
+	req *csi.NodeStageVolumeRequest,
+	disableInUseChecks bool,
+	volID string, cr *util.Credentials) (*rbdVolume, error) {
```
Comment (Madhu-1) [inline internal/rbd/nodeserver.go]: volID is part of req, reusing avoid the number of arguments.

File: internal/rbd/nodeserver.go [hunk 6]
```diff
@@ -149,6 +149,65 @@ func healerStageTransaction(ctx context.Context, cr *util.Credentials, volOps *r
...
+	var err error
+	var j *journal.Connection
+	volOptions, err := genVolFromVolumeOptions(ctx, req.GetVolumeContext(), req.GetSecrets(), disableInUseChecks)
```
Comment (Madhu-1) [inline internal/rbd/nodeserver.go]: `disableInUseChecks` is used only in genVolFromOptions.  we can move the calculation of `disableInUseChecks` to this function it no need to be an argument.

File: internal/rbd/nodeserver.go [hunk 7]
```diff
@@ -149,6 +149,65 @@ func healerStageTransaction(ctx context.Context, cr *util.Credentials, volOps *r
 	return nil
 }
 
+// populateVolOptions update the fields in rbdVolume struct based on the request it received.
+func populateRbdVol(
```
Comment (Madhu-1) [inline internal/rbd/nodeserver.go:153]: `populateVolOptions` to `populateRbdVol`? function name and comment is not matching

**Review by Madhu-1 — DISMISSED** [review]

File: internal/rbd/nodeserver.go (hunk 5 above)
Comment (humblec) [reply internal/rbd/nodeserver.go]: done..

File: internal/rbd/nodeserver.go (hunk 6 above)
Comment (humblec) [reply internal/rbd/nodeserver.go]: Done..

File: internal/rbd/nodeserver.go (hunk 7 above)
Comment (humblec) [reply internal/rbd/nodeserver.go:153]: Corrected. ptal .. thanks @Madhu-1

**Review by Madhu-1 — COMMENTED** [review]
>internal/rbd/nodeserver.go:251: File is not `gofumpt`-ed (gofumpt)

Need to fix CI

**Comment by humblec** [comment]
> internal/rbd/nodeserver.go

@madhu-1 done.. thanks.

**Review by Madhu-1 — DISMISSED** [review]

**Comment by humblec** [comment]
@Mergifyio rebase

File: internal/rbd/nodeserver.go (hunk 1 above)
Comment (nixpanic) [reply internal/rbd/nodeserver.go]: it has not changed, still `volOptions` and not `rbdVol`?

Comment (humblec) [reply internal/rbd/nodeserver.go]: @nixpanic can you refresh and see ?  it is `rbdVol` https://github.com/ceph/ceph-csi/pull/2482/files#diff-2f84daaad04728739d84bfee31afc54913845fde94c8d612079dc36f45bdb5f6R153

**Comment by humblec** [comment]
@nixpanic  all the comment are addressed, ptal

File: internal/rbd/nodeserver.go (hunk 1 above)
Comment (nixpanic) [reply internal/rbd/nodeserver.go]: On line 181 the variable is still confusingly called `volOptions`, not `rv` or `rbdVol`. Refreshing does not seem to be sufficient.

File: internal/rbd/nodeserver.go [hunk 8]
```diff
@@ -280,6 +226,74 @@ func (ns *NodeServer) NodeStageVolume(
...
+
+	isStaticVol := isStaticVolume(req.GetVolumeContext())
+
+	// throw error when imageFeatures parameter is missing or empty
+	// for backward compatibility, ignore error for non-static volumes from older cephcsi version
+	if imageFeatures, ok := req.GetVolumeContext()["imageFeatures"]; checkImageFeatures(imageFeatures, ok, isStaticVol) {
+		return nil, status.Error(codes.InvalidArgument, "missing required parameter imageFeatures")
+	}
+
+	volOptions, err := populateRbdVol(ctx, req, cr)
```
Comment (nixpanic) [inline internal/rbd/nodeserver.go]: `volOptions` here as well, the object does not contain only options, but has a lot of methods too. It is clearer to rename it to `rv` or `rbdVol` as it is not a volume-options object independently of cephfs/rbd (hopefully we get there is in the future, but we're still far away from that).

Comment (humblec) [reply internal/rbd/nodeserver.go]: moved the variable to `rv` , ptal.. thanks @nixpanic

File: internal/rbd/nodeserver.go (hunk 1 above)
Comment (humblec) [reply internal/rbd/nodeserver.go]: changed those too..

File: internal/rbd/nodeserver.go (hunk 8 above)
Comment (humblec) [reply internal/rbd/nodeserver.go]: please wait.. let me rerun the tests.

Comment (humblec) [reply internal/rbd/nodeserver.go]: should be good now;

**Review by nixpanic — APPROVED** [review]
Thanks!

**Review by Madhu-1 — APPROVED** [review]

**Pull request merged**